Merge torchdim into functorch build

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82454

This moves first-class dimensions, as prototyped in https://github.com/facebookresearch/torchdim
into the functorch build. This makes them availiable for use in PrimTorch more easily.

zdevito committed:
Merge torchdim into functorch build

This moves first-class dimensions, as prototyped in https://github.com/facebookresearch/torchdim
into the functorch build. This makes them availiable for use in PrimTorch more easily.

[ghstack-poisoned]

zdevito committed:
Update on "Merge torchdim into functorch build"


This moves first-class dimensions, as prototyped in https://github.com/facebookresearch/torchdim
into the functorch build. This makes them availiable for use in PrimTorch more easily.

[ghstack-poisoned]

zdevito committed:
Update on "Merge torchdim into functorch build"


This moves first-class dimensions, as prototyped in https://github.com/facebookresearch/torchdim
into the functorch build. This makes them availiable for use in PrimTorch more easily.

[ghstack-poisoned]

zou3519 commented on functorch/functorch/dim/__init__.py:
```diff
@@ -0,0 +1,170 @@
+from .tree_map import tree_flatten, tree_map
+from .wrap_type import wrap_type
+from functorch._C import dim as _C
+_C._patch_tensor_class()
```
To check my understanding: the monkey patching for torchdim happens on `import functorch.dim` or `from functorch.dim import ...` and does not happen if someone just calls `import functorch`, right?

zdevito replied:
Correct, I was careful that no patching happens unless the dim package is imported. A follow up is to either write a more in-depth `__torch_function__` for getitem/setitem, or to simply merge dim handling into getitem/setitem code.

zou3519 replied:
(y) thanks for confirming

zou3519 commented:
Test failures are legit

zou3519 commented on functorch/test/test_dims.py:
```diff
@@ -0,0 +1,592 @@
+    print(name, elapsed / r)
+    return elapsed / r
+
+class TestMin(TestCase):
```
For a test to be runnable via the pytorch testing infrastructure (i.e. run_test.py), it needs to extend from torch.testing._internal.common_utils.TestCase. That might just be a drop-in replacement for unittest.TestCase here

zdevito committed:
Update on "Merge torchdim into functorch build"


This moves first-class dimensions, as prototyped in https://github.com/facebookresearch/torchdim
into the functorch build. This makes them availiable for use in PrimTorch more easily.

[ghstack-poisoned]

zdevito committed:
Update on "Merge torchdim into functorch build"


This moves first-class dimensions, as prototyped in https://github.com/facebookresearch/torchdim
into the functorch build. This makes them availiable for use in PrimTorch more easily.

[ghstack-poisoned]

zdevito committed:
Update on "Merge torchdim into functorch build"


This moves first-class dimensions, as prototyped in https://github.com/facebookresearch/torchdim
into the functorch build. This makes them availiable for use in PrimTorch more easily.

[ghstack-poisoned]

zdevito commented:
@pytorchbot merge

zou3519 commented:
@pytorchbot merge

ezyang commented:
@pytorchbot rebase

zdevito committed:
Update on "Merge torchdim into functorch build"


This moves first-class dimensions, as prototyped in https://github.com/facebookresearch/torchdim
into the functorch build. This makes them availiable for use in PrimTorch more easily.

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

zdevito commented:
@pytorchbot rebase

zdevito committed:
Update on "Merge torchdim into functorch build"


This moves first-class dimensions, as prototyped in https://github.com/facebookresearch/torchdim
into the functorch build. This makes them availiable for use in PrimTorch more easily.

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

zengk95 commented:
@pytorchbot revert -m "this is breaking mac jobs on trunk https://hud.pytorch.org/pytorch/pytorch/commit/23b90044dac04d23f43c1a0f518bdbf95efd3b47" -c nosignal

zdevito committed:
Update on "Merge torchdim into functorch build"


This moves first-class dimensions, as prototyped in https://github.com/facebookresearch/torchdim
into the functorch build. This makes them availiable for use in PrimTorch more easily.

[ghstack-poisoned]

zdevito commented:
@pytorchbot rebase

zdevito committed:
Update on "Merge torchdim into functorch build"


This moves first-class dimensions, as prototyped in https://github.com/facebookresearch/torchdim
into the functorch build. This makes them availiable for use in PrimTorch more easily.

[ghstack-poisoned]

zdevito committed:
Update on "Merge torchdim into functorch build"


This moves first-class dimensions, as prototyped in https://github.com/facebookresearch/torchdim
into the functorch build. This makes them availiable for use in PrimTorch more easily.

[ghstack-poisoned]

zdevito commented:
@pytorchbot rebase

zdevito committed:
Update on "Merge torchdim into functorch build"


This moves first-class dimensions, as prototyped in https://github.com/facebookresearch/torchdim
into the functorch build. This makes them availiable for use in PrimTorch more easily.

[ghstack-poisoned]

zdevito committed:
Update on "Merge torchdim into functorch build"


This moves first-class dimensions, as prototyped in https://github.com/facebookresearch/torchdim
into the functorch build. This makes them availiable for use in PrimTorch more easily.

[ghstack-poisoned]

zdevito committed:
Update on "Merge torchdim into functorch build"


This moves first-class dimensions, as prototyped in https://github.com/facebookresearch/torchdim
into the functorch build. This makes them availiable for use in PrimTorch more easily.

[ghstack-poisoned]

zdevito commented:
@pytorchbot merge

swolchok commented on functorch/functorch/csrc/dim/arena.h:
```diff
@@ -0,0 +1,328 @@
+    : impl_(nullptr){}
+    TensorRef(const at::Tensor& t)
+    : impl_(t.unsafeGetTensorImpl()) {}
+    const at::Tensor& operator*() const {
+        return *(at::Tensor*)this;
+    }
+    at::Tensor* operator->() const {
+        return (at::Tensor*)this;
+    }
```
these casts have undefined behavior, and because of TBAA I wouldn't be surprised at all if you were actually bitten by this in practice.

see the aliasing rules at https://en.cppreference.com/w/cpp/language/reinterpret_cast and note that C-style casts are interpreted as whichever fancy_cast matches, which will be `reinterpret_cast` in this case https://en.cppreference.com/w/cpp/language/explicit_cast
